bindings/nodejs/samples/calamari.js

```diff
@@ -0,0 +1,110 @@
+ * limitations under the License.
+ */
+
+/* A simple sample of nodejs Soletta API. On a Calamari Lure - on top of a
+ * MinnowBoard MAX - pressing buttons should toggle each colour of RGB led.
+ * Moving Calamari lever shall also change intensit of lure PWM LED1.
+ */
+
+gp = require('soletta/gpio')
+spi = require('soletta/spi')
+pwm = require('soletta/pwm')
+
+var leverInterval;
+var pwmLed;
+var gpios = [];
+var lever;
+
+gpio_map = [{pin: 338}, {pin: 339}, {pin: 464},
+            {pin: 472, direction: "in", edge: "rising"},
+            {pin: 483, direction: "in", edge: "rising"},
+            {pin: 482, direction: "in", edge: "rising"}]
+            .map(function(pin_config) {
+                return new Promise(function(fullfill) {
+                    gp.open(pin_config)
+                        .then((gpio) => {
+                            gpios.push(gpio);
+                            fullfill(gpio)
+                        })
+                        .catch(function(fail) {
+                            console.log("Could not open gpio: ", fail);
+                            fullfill();
+                        });
+                });
+            });
+
+/* Hack in place: I really don't care about fullfilled values, only about
+ * `gpios` array itself, as it is the one that contains valid gpios.
+ * If we don't have six of them, then application failed. At least, we can
+ * close all gpio that were opened at exit handler */
+Promise.all(gpio_map).then(_ => {
+    if (gpios.length == 6) {
+        btnLedControllers = [
+            {led: gpios[0], btn: gpios[3]},
+            {led: gpios[1], btn: gpios[4]},
+            {led: gpios[2], btn: gpios[5]},
+        ];
+        btnLedControllers.map(function(controller) {
+            var status = {status: true};
+            controller.btn.addEventListener("change", function(status, event) {
+                status.status = !status.status;
+                controller.led.write(status.status);
+            }.bind(this, status));
+        });
+    } else {
+        process.exit();
+    }
+});
+
+spi.open({bus: 0, frequency: 100 * 1000})
+    .then(function(_lever) {
+        lever = _lever;
+        leverInterval = setInterval(function() {
+            lever.transfer([0x01, 0x80, 0x00])
+                .then(function(read) {
+                    var val = (read[1] << 8 | read[2]) & 0x3ff;
+                    if (pwmLed) {
+                        pwmLed.setDutyCycle(Math.floor(val * 10000 / 1023));
+                    }
+                })
+        }, 100)
+    })
+    .catch(function(fail){
+        console.log("Could not open SPI: ", fail);
+        process.exit();
+    });
+
+pwm.open({device: 0, channel: 0, period: 10000, dutyCycle: 0, enabled: true})
+    .then(function(led){
+        pwmLed = led;
+    })
+    .catch(function(fail){
+        console.log("Could not open PWM: ", fail);
+        process.exit();
+    });
+
+process.on("exit", (code) => {
```

Comment:
This is the correct way of closing opened IOs on exit. AFAIK, there is no automatic way of doing it properly, so we leave it up to developers to do the cleanup properly on exit.